### DISCUSSION TITLE
Solved: invoking templater commands through advanced uri may causes slowdows

### DISCUSSION THREAD

**Discussion by imed-ghomari** [Taitava/obsidian-shellcommands#187] [Debugging and testing]
I'm using this templater script to update a counter metadata for some files
```
<%*
const {update, getPropertyValue, getFilesWithProperty} = app.plugins.plugins['metaedit'].api
const file = tp.file.find_tfile(tp.file.title) 
const files = await getFilesWithProperty('Revision')
if (files => files.basename.includes(tp.file.title)) {
    const counter = Number(await getPropertyValue('Revision',file)) + 1
    await update('Revision',counter,file)
}
_%>
```
It is invoked with advanced obsidian uri plugin like this `open "obsidian://advanced-uri?vault=obsidian&commandid=templater-obsidian%253A%2523%2520obsidian%252Ftemplates%252Ftest.md"`

I'm using the `After switching the active pane` event, but now, I'm noticing a slowdown when switching panes. I tested the script using templater only or though advanced uri plugin, and it's very fast.

Likewise, I tried executing the command manually with SC, and still the same delay. What's causing this in your opinion ?

**Comment by Taitava** [comment]
What operating system and shell are you using? [If you don't know the shell, here are instructions how you can find it out](https://publish.obsidian.md/shellcommands/Operating+systems+and+shells/Shells#How+to+know+which+shell+is+used).

If the shell happens to be PowerShell, then there is unfortunately always some delay (1 - 2 seconds) when executing a shell command. I don't know of a way to fix it. But if you use some other shell, then there might be some other reason.

**Reply by imed-ghomari** [reply]
I'm using the default on Mac (bash).

**Reply by Taitava** [reply]
Okay, then it's something that's new to me. I haven't heard before that there would be some latency when using Bash, so I'm quite interested in finding out the reason.

1. Can you try a more simple shell command, e.g. just `echo Hello` and make the output directed to a [notification balloon](https://publish.obsidian.md/shellcommands/Output+handling/Output+channel+-+Notification+balloon)? Then it's possible to see if the latency is somehow related to the particular shell command you used first, or if it happens for all shell commands.

2. If you are willing to use some extra effort, you could try installing [Dash](https://en.wikipedia.org/wiki/Almquist_shell#dash) or [Zsh](https://www.zsh.org/) shell on your Mac and then configure SC to use either of those. Then see if there's difference in how fast or slow the shell command execution is.

3. Do you happen to have your Obsidian vault stored on a network drive?

**Reply by imed-ghomari** [reply]
1. I tried `echo Hello`, the delay is the same.
2. Okay, I'll try the 2 options.
3. No, locally.

**Reply by imed-ghomari** [reply]
Same delay with Dash and Zsh

**Reply by Taitava** [reply]
I'll try to come up with another thing that you can test, but I'm in a bit of hurry now, so I'll write about it later.

Meanwhile, you can try to disable all other plugins and only leave SC enabled. Then try to execute the `echo Hello` shell command again and see if it still has the same delay. I guess disabling other plugins won't help at all, but at least it can be tried.

**Reply by imed-ghomari** [reply]
The delay is still there when only SC in enabled.

**Reply by Taitava** [reply]
Ok, now to the idea I had earlier today: Can you use Templater to execute the same `echo Hello` shell command?

You can define it like this in Templater's settings:

![kuva](https://user-images.githubusercontent.com/13002142/159965858-a1f781d7-d935-455c-a145-30d2f3005f87.png)

And call it like this:

```
<%+ tp.user.echotest() %>
```

What I'm interested to find out, is that when you switch to preview mode, is there still some delay, or will the _Hello_ result appear immediately? If it appears immediately, then I assume there's something wrong with SC, but if also this test has delay, then I assume the problem might be somewhere out of the scope of SC.

---

I also have another thing in my mind that might work for you. It doesn't fix the actual problem, however.

I'm not very familiar with Templater, but now that I used it quickly, I noticed that it has a feature called _Startup Templates_, that you can use to hook into Obsidian events, such as the [`active-leaf-change` Obsidian event](https://github.com/obsidianmd/obsidian-api/blob/763a243b4ec295c9c460560e9b227c8f18d8199b/obsidian.d.ts#L3576) that SC's _After switching the active pane_ uses internally. Maybe you could hook into that event and have your Templater script run that way every time you switch between panes? Then you would not need SC nor Advanced Obsidian URI for this particular thing.

I'm don't have the best knowledge on how to make this work exactly, but if you need help, maybe you can ask in the [Templater's Help discussion section](https://github.com/SilentVoid13/Templater/discussions/categories/help), or in [Obsidian's Discord server](https://discord.gg/veuWUTm).

---

That being said, we can still try to find out if there's anything to fix in SC to make the delay go away. 🙂

**Reply by imed-ghomari** [reply]
When trying `<%+ tp.user.echotest() %>`, there is still a delay, so it appears it's out of the scope of SC, sorry.

Thank you for the workaround.

**Reply by Taitava** [reply]
No problem!

Sad that we were not able to find out the exact reason for the delay 🙁. But I hope the workaround works for you! 🙂

**Reply by imed-ghomari** [reply]
Just wanted to let you know that there is no more lag now, and I guess it was the theme I used at the time that caused it.

**Reply by Taitava** [reply]
Okay, thanks for letting me know! 🙂 Strange how a theme can cause a lag like this. If I have understood correctly, Obsidian's themes consists of just CSS? But anyway, nice to know that you were able to solve this! 😎
